mqtt/mqtt-bridge/src/persist/storage/ring_buffer/mod.rs

```diff
@@ -0,0 +1,693 @@
+
+            // to move to beginning of next block header
+            // start += 1;
+            validate(&block, &data)?;
+            let publication = binary_deserialize::<Publication>(data.inner())?;
+            let key = Key {
+                offset: index as u64,
+            };
+            vdata.push_back((key, publication));
+        }
+
+        Ok(vdata)
+    }
+
+    fn remove(&mut self, key: usize) -> StorageResult<()> {
+        let timer = Instant::now();
+        let read_index = self.pointers.read.load(Ordering::SeqCst);
+        if key != read_index {
+            return Err(StorageError::RingBuffer(RingBufferError::RemovalIndex));
+        }
+        let block_size = serialized_block_size()?;
+
+        let start = key;
+        let mut end = start + block_size;
+        if end > self.max_file_size {
+            return Err(StorageError::RingBuffer(RingBufferError::WrapAround.into()));
+        }
+
+        let maybe_block = deserialize_block(self.inner.clone(), start, end);
+        if maybe_block.is_err() {
+            return Err(StorageError::RingBuffer(RingBufferError::NonExistantKey));
+        }
+
+        let data_size;
+        let mut block = maybe_block.unwrap();
+        if block.inner().hint() != BLOCK_HINT {
+            return Err(StorageError::RingBuffer(RingBufferError::NonExistantKey));
+        }
+
+        {
+            let inner_block = block.inner_mut();
+            inner_block.set_should_not_overwrite(false);
+            data_size = inner_block.data_size();
+        }
+
+        let bytes = binary_serialize(&block)?;
+        let should_flush = should_flush(self.flush_options, &self.flush_state);
+        mmap_write(self.inner.clone(), start, end, &bytes, should_flush)?;
+
+        end += data_size;
+
+        self.pointers
+            .read
+            .store(end % self.max_file_size, Ordering::SeqCst);
+
+        self.flush_state
+            .update(0, 0, timer.elapsed().as_millis() as usize);
+        if should_flush {
+            self.flush_state.reset(&self.flush_options);
+        }
+
+        Ok(())
+    }
+
+    fn set_waker(&mut self, waker: &Waker) {
+        self.waker = Some(waker.clone());
+    }
+
+    fn wake_up_task(&mut self) {
+        if let Some(waker) = self.waker.take() {
+            waker.wake();
+        }
+    }
+}
+
+#[allow(dead_code)]
+impl Drop for RingBuffer {
+    fn drop(&mut self) {
+        let mmap = self.inner.lock();
+        mmap.flush().expect("Failed to flush");
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use bytes::Bytes;
+    use matches::assert_matches;
+    use mqtt3::proto::QoS;
+    use rand::{distributions::Alphanumeric, thread_rng, Rng};
+    use std::{
+        fs::{remove_dir_all, remove_file},
+        panic,
+    };
+
+    const FLUSH_OPTIONS: FlushOptions = FlushOptions::Off;
+    const FILE_NAME: &'static str = "test_file";
+    const MAX_FILE_SIZE: usize = 1024;
+
+    fn cleanup_test_file(file_name: String) {
```

Comment:
we use `tempfile` crate for these temporary files

Reply:
Oh nice! This is exactly what I need!